commit 1acecf80b4623dc7e5916c8376d469ecc0e3f5a1
Author: Phil Elwell <phil@raspberrypi.org>
Date:   Fri Oct 26 17:29:51 2018 +0100

    mmc/bcm2835-sdhost: Recover from MMC_SEND_EXT_CSD
    
    If the user issues an "mmc extcsd read", the SD controller receives
    what it thinks is a SEND_IF_COND command with an unexpected data block.
    The resulting operations leave the FSM stuck in READWAIT, a state which
    persists until the MMC framework resets the controller, by which point
    the root filesystem is likely to have been unmounted.
    
    A less heavyweight solution is to detect the condition and nudge the
    FSM by asserting the (self-clearing) FORCE_DATA_MODE bit.
    
    N.B. This workaround was essentially discovered by accident and without
    a full understanding the inner workings of the controller, so it is
    fortunate that the "fix" only modifies error paths.
    
    See: https://github.com/raspberrypi/linux/issues/2728
    
    Signed-off-by: Phil Elwell <phil@raspberrypi.org>

diff --git a/drivers/mmc/host/bcm2835-sdhost.c b/drivers/mmc/host/bcm2835-sdhost.c
index 34ae2be3647f..449194d3c271 100644
--- a/drivers/mmc/host/bcm2835-sdhost.c
+++ b/drivers/mmc/host/bcm2835-sdhost.c
@@ -1244,6 +1244,8 @@ static void bcm2835_sdhost_finish_command(struct bcm2835_host *host,
 				pr_info("%s: ignoring CRC7 error for CMD1\n",
 					mmc_hostname(host->mmc));
 		} else {
+			u32 edm, fsm;
+
 			if (sdhsts & SDHSTS_CMD_TIME_OUT) {
 				if (host->debug)
 					pr_warn("%s: command %d timeout\n",
@@ -1256,6 +1258,13 @@ static void bcm2835_sdhost_finish_command(struct bcm2835_host *host,
 				       host->cmd->opcode);
 				host->cmd->error = -EILSEQ;
 			}
+
+			edm = readl(host->ioaddr + SDEDM);
+			fsm = edm & SDEDM_FSM_MASK;
+			if (fsm == SDEDM_FSM_READWAIT ||
+			    fsm == SDEDM_FSM_WRITESTART1)
+				writel(edm | SDEDM_FORCE_DATA_MODE,
+				       host->ioaddr + SDEDM);
 			tasklet_schedule(&host->finish_tasklet);
 			return;
 		}
diff -uN a/1.txt b/1.txt
--- a/dummy/rpi_1649_1acecf80b4623dc7e5916c8376d469ecc0e3f5a1.txt	1970-01-01 00:00:00.000000000 +0000
+++ b/dummy/rpi_1649_1acecf80b4623dc7e5916c8376d469ecc0e3f5a1.txt	2013-12-23 04:07:40.000000000 +0000
@@ -0,0 +1 @@
+dummy file to ensure patch has content.
